pkg/server/tls.go

```diff
@@ -0,0 +1,85 @@
+package server
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"io/ioutil"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	clientCertExpiry = newLazyGauge(prometheus.GaugeOpts{
+		Name: "grpc_client_tls_certificate_expiry_seconds",
+		Help: "Expiration time of the gRPC client TLS certificate in seconds since the Unix epoch.",
+	})
+	serverCertExpiry = newLazyGauge(prometheus.GaugeOpts{
+		Name: "grpc_server_tls_certificate_expiry_seconds",
+		Help: "Expiration time of the gRPC server TLS certificate in seconds since the Unix epoch.",
+	})
+)
+
+type lazyGauge struct {
+	init   sync.Once
+	metric prometheus.Gauge
+}
+
+func newLazyGauge(opts prometheus.GaugeOpts) *lazyGauge {
+	return &lazyGauge{metric: prometheus.NewGauge(opts)}
+}
+
+func (g *lazyGauge) Set(v float64) {
+	g.init.Do(func() { prometheus.MustRegister(g.metric) })
+	g.metric.Set(v)
+}
```

Comment:
what is the reason for creating this wrapper around gauge? I'd rather keep things simple and readable instead of creating a custom struct for the sake of saving a few lines here and there

Reply:
I only wanted to register the metrics when they are used (e.g. I didn't want the client metric to exist on the server and vice versa).

Reply:
I added more metrics that are symmetric (server and client), so I moved them into a struct that self-registers.